applications/FluidDynamicsApplication/custom_elements/fluid_adjoint_element.cpp

```diff
@@ -0,0 +1,717 @@
+    IndexType NewId,
+    GeometryType::Pointer pGeometry)
+    : BaseType(NewId, pGeometry)
+{
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::FluidAdjointElement(
+    IndexType NewId,
+    GeometryType::Pointer pGeometry,
+    PropertiesType::Pointer pProperties)
+    : BaseType(NewId, pGeometry, pProperties)
+{
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::~FluidAdjointElement()
+{
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+Element::Pointer FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::Create(
+    IndexType NewId,
+    NodesArrayType const& ThisNodes,
+    PropertiesType::Pointer pProperties) const
+{
+    KRATOS_TRY
+    return Kratos::make_intrusive<FluidAdjointElement>(
+        NewId, Element::GetGeometry().Create(ThisNodes), pProperties);
+    KRATOS_CATCH("");
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+Element::Pointer FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::Create(
+    IndexType NewId,
+    GeometryType::Pointer pGeom,
+    PropertiesType::Pointer pProperties) const
+{
+    KRATOS_TRY
+    return Kratos::make_intrusive<FluidAdjointElement>(NewId, pGeom, pProperties);
+    KRATOS_CATCH("");
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+Element::Pointer FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::Clone(
+    IndexType NewId,
+    NodesArrayType const& ThisNodes) const
+{
+    KRATOS_TRY
+    return Kratos::make_intrusive<FluidAdjointElement>(
+        NewId, Element::GetGeometry().Create(ThisNodes), Element::pGetProperties());
+    KRATOS_CATCH("");
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+int FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::Check(const ProcessInfo& rCurrentProcessInfo) const
+{
+    TAdjointElementData::Check(*this, rCurrentProcessInfo);
+    return 0;
+}
+
+template <unsigned int TDim, unsigned int TNumNodes, class TAdjointElementData>
+void FluidAdjointElement<TDim, TNumNodes, TAdjointElementData>::EquationIdVector(
+    EquationIdVectorType& rElementalEquationIdList,
+    const ProcessInfo& rCurrentProcessInfo) const
+{
+    if (rElementalEquationIdList.size() != TElementLocalSize) {
+        rElementalEquationIdList.resize(TElementLocalSize, false);
+    }
+
+    const auto& r_variables_list = TAdjointElementData::GetDofVariablesList();
+
+    IndexType local_index = 0;
+    for (IndexType i = 0; i < TNumNodes; ++i) {
+        const auto& r_node = this->GetGeometry()[i];
+        for (const auto p_variable : r_variables_list) {
+            rElementalEquationIdList[local_index++] =
+                r_node.GetDof(*p_variable).EquationId();
```

Comment:
Use the access with position, it's much more efficient.
```c++
    const unsigned int den_pos = r_geometry[0].GetDofPosition(DENSITY);
    const unsigned int mom_pos = r_geometry[0].GetDofPosition(MOMENTUM);
    const unsigned int enr_pos = r_geometry[0].GetDofPosition(TOTAL_ENERGY);
    for (unsigned int i_node = 0; i_node < n_nodes; ++i_node) {
        rResult[local_index++] = r_geometry[i_node].GetDof(DENSITY, den_pos).EquationId();
        rResult[local_index++] = r_geometry[i_node].GetDof(MOMENTUM_X, mom_pos).EquationId();
        rResult[local_index++] = r_geometry[i_node].GetDof(MOMENTUM_Y, mom_pos + 1).EquationId();
        rResult[local_index++] = r_geometry[i_node].GetDof(TOTAL_ENERGY, enr_pos).EquationId();
    }
```

Reply:
I'm taking it via a list of DOFs, so I am not sure how to use position index in this case. :/

```c++
    // this list is taken first for dof list
    const auto& r_variables_list = TAdjointElementData::GetDofVariablesList();

    IndexType local_index = 0;
    for (IndexType i = 0; i < TNumNodes; ++i) {
        const auto& r_node = this->GetGeometry()[i];
        for (const auto p_variable : r_variables_list) {
            rElementalEquationIdList[local_index++] =
                r_node.GetDof(*p_variable).EquationId();
        }
    }
```

Reply:
This is still pending. We should use the DOF with position access.